registry-plugin/main.go

```diff
@@ -0,0 +1,246 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"google.golang.org/grpc/metadata"
+
+	mgmtPB "github.com/instill-ai/protogen-go/core/mgmt/v1beta"
+)
+
+// pluginName is the plugin name
+var pluginName = "registry"
+
+// HandlerRegisterer is the symbol the plugin loader will try to load. It must implement the Registerer interface
+var HandlerRegisterer = registerer(pluginName)
+
+type registerer string
+
+func (r registerer) RegisterHandlers(f func(
+	name string,
+	handler func(context.Context, map[string]interface{}, http.Handler) (http.Handler, error),
+)) {
+	f(string(r), r.registerHandlers)
+}
+
+func (r registerer) registerHandlers(ctx context.Context, extra map[string]interface{}, h http.Handler) (http.Handler, error) {
+
+	config, ok := extra[pluginName].(map[string]interface{})
+	if !ok {
+		return h, errors.New("configuration not found")
+	}
+
+	hostport, _ := config["hostport"].(string)
+
+	mgmtPublicClient, _ := InitMgmtPublicServiceClient(ctx, config["mgmt_public_hostport"].(string), "", "")
+	mgmtPrivateClient, _ := InitMgmtPrivateServiceClient(ctx, config["mgmt_private_hostport"].(string), "", "")
+
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+
+		// If the URL path starts with "/v2/" (exactly /v2/ indicating the first handshake request to confirm registry V2 API),
+		// it means that the request is intended for the Instill Artifact registry. In this case, the traffic is hijacked
+		// and directly relayed to the registry. Otherwise, if the URL path does not match any of these patterns,
+		// the traffic is passed through to the next handler.
+		if !strings.HasPrefix(req.URL.Path, "/v2/") {
+			h.ServeHTTP(w, req)
+			return
+		}
+
+		// Authenticate the user via docker login
+		var username, password string
+		if req.URL.Path == "/v2/" {
+			username, password, ok = req.BasicAuth()
+			if !ok {
+				// Challenge the user for basic authentication
+				w.Header().Set("WWW-Authenticate", `Basic realm="Restricted"`)
+				writeStatusUnauthorized(req, w, "Instill AI user authentication failed")
+				return
+			}
+
+			// Validate the api key and the namespace authorization
+			if !strings.HasPrefix(password, "instill_sk_") {
+				writeStatusUnauthorized(req, w, "Instill AI user authentication failed")
+				return
+			}
+
+			ctx := metadata.AppendToOutgoingContext(ctx, "Authorization", fmt.Sprintf("Bearer %s", password))
+			if resp, err := mgmtPublicClient.ValidateToken(
+				ctx,
+				&mgmtPB.ValidateTokenRequest{},
+			); err == nil {
```

Comment:
Better handle the "unauthorized" or other unhappy paths eariler. Reduces the indentation and the brain tends to see "if _, err := f(params); err == nil {" as an error comparison (out of habit). In this case it prevents us from reusing the `resp` variable, but that isn't necessarily a bad thing.

```go
		if req.URL.Path == "/v2/" {
			username, password, ok = req.BasicAuth()
			if !ok {
				// Challenge the user for basic authentication
				w.Header().Set("WWW-Authenticate", `Basic realm="Restricted"`)
				writeStatusUnauthorized(req, w, "Instill AI user authentication failed")
				return
			}

			// Validate the api key and the namespace authorization
			if !strings.HasPrefix(password, "instill_sk_") {
				writeStatusUnauthorized(req, w, "Instill AI user authentication failed")
				return
			}

			ctx := metadata.AppendToOutgoingContext(ctx, "Authorization", fmt.Sprintf("Bearer %s", password))
			tokenValidation, err := mgmtPublicClient.ValidateToken(
				ctx,
				&mgmtPB.ValidateTokenRequest{},
			)
			if err != nil {
				writeStatusInternalError(req, w)
				return
			}

			userUID := tokenValidation.UserUid
			// Check if the login username is the same with the user id retrieved from the token validation response
			userLookup, err := mgmtPrivateClient.LookUpUserAdmin(
				ctx,
				&mgmtPB.LookUpUserAdminRequest{
					Permalink: "users/" + userUID,
				},
			)
			if err != nil {
				logger.Error(err.Error())
				writeStatusInternalError(req, w)
				return
			}

			if userLookup.User.Id != username {
				writeStatusUnauthorized(req, w, "Instill AI user authentication failed")
				return
			}

			// To this point, if the req.URL.Path is "/v2/", return 200 OK to the client for login success
			writeStatusOK(req, w)
			return
		}
```